src/controllers/secret.go

```diff
@@ -0,0 +1,205 @@
+	for _, secret := range secrets.Items {
+		response.Secrets = append(
+			response.Secrets,
+			types.Secret{
+				Type:      string(secret.Type),
+				Name:      secret.Name,
+				Namespace: secret.Namespace,
+			})
+	}
+	return response, nil
+}
+
+func (n *secretController) GetSecret(namespace string, name string) (types.GetSecretResponse, error) {
+	n.logger.Debug(fmt.Sprintf("Trying to get %q secret in %q namespace", name, namespace))
+
+	response := types.GetSecretResponse{}
+	secret, err := n.client.CoreV1().Secrets(namespace).Get(n.ctx, name, metav1.GetOptions{})
+	if err != nil {
+		n.logger.Error(fmt.Sprintf("Could not get secret %q with error: %v", name, err.Error()))
+		return response, err
+	}
+
+	n.logger.Debug("Fetched secret successfully")
+
+	response.Id = string(secret.GetUID())
+	response.Name = secret.Name
+	response.Type = string(secret.Type)
+	for k, v := range secret.Data {
+		response.Data = append(response.Data, types.KeyValue{
+			Key:   k,
+			Value: string(v),
+		})
+	}
+	return response, nil
+}
+
+func (n *secretController) PatchSecret(namespace string, name string, request types.PatchSecretRequest) (types.PatchSecretResponse, error) {
+	n.logger.Debug(fmt.Sprintf("Trying to patch an existing secret in %q namespace", namespace))
+
+	response := types.PatchSecretResponse{}
+	secret := v1.Secret{
+		Data: map[string][]byte{},
+	}
+	for _, kv := range request.Data {
+		secret.Data[kv.Key] = []byte(kv.Value)
+	}
+
+	data, err := json.Marshal(secret)
+	if err != nil {
+		n.logger.Error(fmt.Sprintf("Could not marshal secret %q with error: %v", name, err.Error()))
+		return response, err
+	}
+
+	result, err := n.client.CoreV1().Secrets(namespace).Patch(n.ctx, name, k8stypes.StrategicMergePatchType, data, metav1.PatchOptions{})
+	if err != nil {
+		n.logger.Error(fmt.Sprintf("Could not patch secret %q with error: %v", name, err.Error()))
+		return response, err
+	}
+
+	response.Id = string(result.UID)
+	response.Type = string(result.Type)
+	response.Name = result.Name
+	for k, v := range secret.Data {
+		response.Data = append(response.Data, types.KeyValue{
+			Key:   k,
+			Value: string(v),
+		})
+	}
+	return response, nil
+}
+
+func (n *secretController) DeleteSecret(namespace string, name string) (types.DeleteSecretResponse, error) {
+	n.logger.Debug(fmt.Sprintf("Trying to delete an existing secret in %q namespace", namespace))
+
+	response := types.DeleteSecretResponse{}
+	if err := n.client.CoreV1().Secrets(namespace).Delete(n.ctx, name, metav1.DeleteOptions{}); err != nil {
+		n.logger.Error(fmt.Sprintf("Could not delete secret %q with error: %v", name, err.Error()))
+		return response, err
+	}
+
+	response.Message = fmt.Sprintf("Secret %q was deleted successfully", name)
+	return response, nil
+}
+
+// createSecretFromRequest returns a new secret based on different secret
+// types, either TLS or Opaque.
+func createSecretFromRequest(namespace string, request types.CreateSecretRequest) (*v1.Secret, error) {
```

Comment:
Because we have already function called `CreateSecret`  this function name is a bit confusing ill change it to `prepareSecret`

Reply:
I'll change it to `newSecretFromRequest`.